File: pkg/types/workflow/execution.go
package workflow

import "time"

// --- Execution Result & Status ---

type ExecutionStatus string

const (
	ExecutionStatusPending   ExecutionStatus = "pending"
	ExecutionStatusRunning   ExecutionStatus = "running"
	ExecutionStatusSuccess   ExecutionStatus = "success"
	ExecutionStatusError     ExecutionStatus = "error"
	ExecutionStatusPaused    ExecutionStatus = "paused"
	ExecutionStatusCancelled ExecutionStatus = "cancelled"
)

type ExecutionResult struct {
	WorkflowID    string                `json:"workflowId"`
	ExecutionID   string                `json:"executionId"`
	Status        ExecutionStatus       `json:"status"`
	StartTime     time.Time             `json:"startTime"`
	EndTime       *time.Time            `json:"endTime,omitempty"`
	Duration      *float64              `json:"duration,omitempty"` // in seconds
	ExecutedNodes []string              `json:"executedNodes"`
	NodeResults   map[string]NodeResult `json:"nodeResults"`
	Errors        []ExecutionError      `json:"errors"`
	Output        any                   `json:"output"`
	TriggerData   any                   `json:"triggerData,omitempty"`
}

type NodeExecutionStatus string

const (
	NodeExecutionStatusPending   NodeExecutionStatus = "pending"
	NodeExecutionStatusRunning   NodeExecutionStatus = "running"
	NodeExecutionStatusSuccess   NodeExecutionStatus = "success"
	NodeExecutionStatusError     NodeExecutionStatus = "error"
	NodeExecutionStatusSkipped   NodeExecutionStatus = "skipped"
	NodeExecutionStatusRetrying  NodeExecutionStatus = "retrying"
)

type NodeResult struct {
	NodeID      string              `json:"nodeId"`
	Status      NodeExecutionStatus `json:"status"`
	StartTime   time.Time           `json:"startTime"`
	EndTime     *time.Time          `json:"endTime,omitempty"`
	Duration    *float64            `json:"duration,omitempty"` // in seconds
	Input       any                 `json:"input"`
	Output      any                 `json:"output"`
	Error       *NodeError          `json:"error,omitempty"`
	RetryCount  *int                `json:"retryCount,omitempty"`
}

// --- Error Types ---

type ErrorType string

const (
	ErrorTypeValidationError          ErrorType = "validation_error"
	ErrorTypeRuntimeError             ErrorType = "runtime_error"
	ErrorTypeNetworkError             ErrorType = "network_error"
	ErrorTypeAuthenticationError      ErrorType = "authentication_error"
	ErrorTypeTimeoutError             ErrorType = "timeout_error"
	ErrorTypeConfigurationError       ErrorType = "configuration_error"
	ErrorTypeDataTransformationError  ErrorType = "data_transformation_error"
	ErrorTypeWorkflowError            ErrorType = "workflow_error"
)

type NodeErrorType string

const (
	NodeErrorTypeInvalidConfiguration      NodeErrorType = "invalid_configuration"
	NodeErrorTypeMissingRequiredField    NodeErrorType = "missing_required_field"
	NodeErrorTypeAPIRequestFailed        NodeErrorType = "api_request_failed"
	NodeErrorTypeWebhookProcessingFailed NodeErrorType = "webhook_processing_failed"
	NodeErrorTypeConditionEvalFailed     NodeErrorType = "condition_evaluation_failed"
	NodeErrorTypeDataTransformFailed     NodeErrorType = "data_transformation_failed"
	NodeErrorTypeTimerExecFailed         NodeErrorType = "timer_execution_failed"
	NodeErrorTypeLoopExecFailed          NodeErrorType = "loop_execution_failed"
)

type ExecutionError struct {
	ID        string    `json:"id"`
	Type      ErrorType `json:"type"`
	Message   string    `json:"message"`
	Details   any       `json:"details,omitempty"`
	NodeID    *string   `json:"nodeId,omitempty"`
	Timestamp time.Time `json:"timestamp"`
	Stack     *string   `json:"stack,omitempty"`
}

type NodeError struct {
	Type      NodeErrorType `json:"type"`
	Message   string        `json:"message"`
	Details   any           `json:"details,omitempty"`
	Timestamp time.Time     `json:"timestamp"`
	Stack     *string       `json:"stack,omitempty"`
}

// --- Execution Context ---

type ExecutionEnvironment string

const (
	ExecutionEnvironmentDevelopment ExecutionEnvironment = "development"
	ExecutionEnvironmentStaging     ExecutionEnvironment = "staging"
	ExecutionEnvironmentProduction  ExecutionEnvironment = "production"
)

type ExecutionMetadata struct {
	StartTime     time.Time            `json:"startTime"`
	TriggerType   string               `json:"triggerType"`
	TriggerData   any                  `json:"triggerData"`
	UserID        *string              `json:"userId,omitempty"`
	Environment   ExecutionEnvironment `json:"environment"`
}

type ExecutionContext struct {
	WorkflowID    string            `json:"workflowId"`
	ExecutionID   string            `json:"executionId"`
	CurrentNodeID string            `json:"currentNodeId"`
	Data          map[string]any    `json:"data"`
	Variables     map[string]any    `json:"variables"`
	Metadata      ExecutionMetadata `json:"metadata"`
}

// --- Validation Result ---

type ValidationType string

const (
	ValidationTypeRequiredFieldMissing ValidationType = "required_field_missing"
	ValidationTypeInvalidFormat      ValidationType = "invalid_format"
	ValidationTypeInvalidType        ValidationType = "invalid_type"
	ValidationTypeCircularDependency ValidationType = "circular_dependency"
	ValidationTypeOrphanedNode       ValidationType = "orphaned_node"
	ValidationTypeInvalidConnection  ValidationType = "invalid_connection"
	ValidationTypeConfigConflict     ValidationType = "configuration_conflict"
)

type ValidationError struct {
	Type    ValidationType `json:"type"`
	Message string         `json:"message"`
	Field   *string        `json:"field,omitempty"`
	NodeID  *string        `json:"nodeId,omitempty"`
}

type ValidationWarning struct {
	Type    ValidationType `json:"type"`
	Message string         `json:"message"`
	Field   *string        `json:"field,omitempty"`
	NodeID  *string        `json:"nodeId,omitempty"`
}

type ValidationResult struct {
	IsValid  bool                `json:"isValid"`
	Errors   []ValidationError   `json:"errors"`
	Warnings []ValidationWarning `json:"warnings"`
}

File: frontend/lib/types/workflow/execution.ts
/**
 * Execution result and error type definations
 */

// Export result interface
export interface ExecutionResult {
    workflowId: string
    executionId: string
    status: ExecutionStatus
    startTime: Date
    endTime?: Date
    duration?: number
    executedNodes: string[]
    nodeResults: Record<string, NodeResult>
    errors: ExecutionError[]
    output: unknown
    triggerData?: unknown
}

export enum ExecutionStatus {
    PENDING = 'pending',
    RUNNING = 'running',
    SUCCESS = 'success',
    ERROR = 'error',
    PAUSED = 'paused',
    CANCELLED = 'cancelled'
}

// Node execution result interface 
export interface NodeResult {
    nodeId: string
    status: NodeExecutionStatus
    startTime: Date
    endTime?: Date
    duration?: number
    input: unknown
    output: unknown
    error?: NodeError
    retryCount?: number
}

export enum NodeExecutionStatus {
    PENDING = 'pending',
    RUNNING = 'running',
    SUCCESS = 'success',
    ERROR = 'error',
    SKIPPED = 'skipped',
    RETRYING = 'retrying'
}

// Error interfaces
export interface ExecutionError {
    id: string;
    type: ErrorType;
    message: string;
    details?: unknown;
    nodeId?: string;
    timestamp: Date;
    stack?: string;
}

export interface NodeError {
    type: NodeErrorType;
    message: string;
    details?: unknown;
    timestamp: Date;
    stack?: string;
}

export enum ErrorType {
    VALIDATION_ERROR = 'validation_error',
    RUNTIME_ERROR = 'runtime_error',
    NETWORK_ERROR = 'network_error',
    AUTHENTICATION_ERROR = 'authentication_error',
    TIMEOUT_ERROR = 'timeout_error',
    CONFIGURATION_ERROR = 'configuration_error',
    DATA_TRANSFORMATION_ERROR = 'data_transformation_error',
    WORKFLOW_ERROR = 'workflow_error'
}

export enum NodeErrorType {
    INVALID_CONFIGURATION = 'invalid_configuration',
    MISSING_REQUIRED_FIELD = 'missing_required_field',
    API_REQUEST_FAILED = 'api_request_failed',
    WEBHOOK_PROCESSING_FAILED = 'webhook_processing_failed',
    CONDITION_EVALUATION_FAILED = 'condition_evaluation_failed',
    DATA_TRANSFORMATION_FAILED = 'data_transformation_failed',
    TIMER_EXECUTION_FAILED = 'timer_execution_failed',
    LOOP_EXECUTION_FAILED = 'loop_execution_failed'
}

// Execution context interface
export interface ExecutionContext {
    workflowId: string;
    executionId: string;
    currentNodeId: string;
    data: Record<string, unknown>;
    variables: Record<string, unknown>;
    metadata: ExecutionMetadata;
}

export interface ExecutionMetadata {
    startTime: Date;
    triggerType: string;
    triggerData: unknown;
    userId?: string;
    environment: ExecutionEnvironment;
}

export enum ExecutionEnvironment {
    DEVELOPMENT = 'development',
    STAGING = 'staging',
    PRODUCTION = 'production'
}

// Validation result interface
export interface ValidationResult {
    isValid: boolean;
    errors: ValidationError[];
    warnings: ValidationWarning[];
}

export interface ValidationError {
    type: ValidationType;
    message: string;
    field?: string;
    nodeId?: string;
}

export interface ValidationWarning {
    type: ValidationType;
    message: string;
    field?: string;
    nodeId?: string;
}

export enum ValidationType {
    REQUIRED_FIELD_MISSING = 'required_field_missing',
    INVALID_FORMAT = 'invalid_format',
    INVALID_TYPE = 'invalid_type',
    CIRCULAR_DEPENDENCY = 'circular_dependency',
    ORPHANED_NODE = 'orphaned_node',
    INVALID_CONNECTION = 'invalid_connection',
    CONFIGURATION_CONFLICT = 'configuration_conflict'
}